### PR TITLE
[Slider] Add inkColor, thumbColorForState:, trackFillColorForState:

### PR DESCRIPTION
**Pull request by romoore**
Instead of a single `color` property that maps back to multiple properties in
MDCThumbTrack, MDCSlider should make each of these properties explicit.  This
includes:

* Ink color
* Thumb color
* Track fill color
* Track background color

For `thumbColor`, `trackFillColor`, and `trackBackgroundColor`, the values are adjustable per-state
(e.g., for `.disabled`).

Partially implements #3137
Pivotal link: https://www.pivotaltracker.com/story/show/155525171

### CODE REVIEW

File: components/Slider/src/MDCSlider.m [hunk 1]
```diff
@@ -84,37 +88,76 @@ - (void)commonMDCSliderInit {
...
 }
 
-#pragma mark - ThumbTrack passthrough methods
+#pragma mark - Color customization methods
 
-- (void)setTrackBackgroundColor:(UIColor *)trackBackgroundColor {
-  _thumbTrack.trackOffColor =
-      trackBackgroundColor ? trackBackgroundColor : [[self class] defaultTrackOffColor];
+- (void)setTrackFillColor:(UIColor *)fillColor forState:(UIControlState)state {
+  _trackFillColorsForState[@(state)] = fillColor;
```
Comment (jverkoey) [inline components/Slider/src/MDCSlider.m]: If fillColor is nil I believe this will crash.

File: components/Slider/src/MDCSlider.m [hunk 2]
```diff
@@ -84,37 +88,76 @@ - (void)commonMDCSliderInit {
...
+  return _thumbTrack.primaryColor;
 }
 
-- (void)setDisabledColor:(UIColor *)disabledColor {
-  _thumbTrack.trackDisabledColor =
-      disabledColor ?: [[self class] defaultDisabledColor];
-  _thumbTrack.thumbDisabledColor =
-      disabledColor ?: [[self class] defaultDisabledColor];
+- (void)setThumbColor:(UIColor *)thumbColor forState:(UIControlState)state {
+  _thumbColorsForState[@(state)] = thumbColor;
```
Comment (jverkoey) [inline components/Slider/src/MDCSlider.m]: Nil case.

**Review by jverkoey — DISMISSED** [review]

File: components/Slider/src/MDCSlider.m (hunk 1 above)
Comment (romoore) [reply components/Slider/src/MDCSlider.m]: Keys cannot be nil but values can be (it removes the mapping from the dictionary).

```objectivec
NSMutableDictionary *dict = [@{} mutableCopy];
  dict[@(1)] = @"foo";
  NSLog(@"%@", dict); // { 1 = foo; }
  dict[@(1)] = nil;
  dict[@(2)] = nil;
  NSLog(@"%@", dict); // {}
```

File: components/Slider/src/MDCSlider.h [hunk 3]
```diff
@@ -41,29 +41,85 @@
 IB_DESIGNABLE
 @interface MDCSlider : UIControl <NSSecureCoding>
 
+/** When @c YES, the new forState: APIs are enabled. Defaults to @c NO. */
```
Comment (jverkoey) [inline components/Slider/src/MDCSlider.h]: [minornit]: "the forState:"*

This API will eventually not be new :P

File: components/Slider/src/MDCSlider.h [hunk 4]
```diff
@@ -41,29 +41,85 @@
...
+@property(nonatomic, assign, getter=isStatefulAPIEnabled) BOOL statefulAPIEnabled;
+
 /** The delegate for the slider. */
 @property(nullable, nonatomic, weak) id<MDCSliderDelegate> delegate;
 
 /**
- The color of the cursor (thumb) and filled in portion of the track (left side).
+ Sets the color of the thumb to use for the specified state.
```
Comment (jverkoey) [inline components/Slider/src/MDCSlider.h]: [minornit]:

> "Sets the color of the thumb for the specified state."

Drop the "to use" part.

here and throughout.

File: components/Slider/src/MDCSlider.m [hunk 5]
```diff
@@ -84,39 +88,105 @@ - (void)commonMDCSliderInit {
...
+    color = _thumbColorsForState[@(UIControlStateNormal)];
+  }
+  return color;
+}
+
+- (void)updateColorsForState {
+  if (!self.isStatefulAPIEnabled) {
+    return;
+  }
+  NSLog(@"Update colors for state");
```
Comment (jverkoey) [inline components/Slider/src/MDCSlider.m]: Remove the log.

**Review by jverkoey — DISMISSED** [review]

File: components/Slider/src/MDCSlider.h (hunk 3 above)
Comment (romoore) [reply components/Slider/src/MDCSlider.h]: Done

File: components/Slider/src/MDCSlider.h (hunk 4 above)
Comment (romoore) [reply components/Slider/src/MDCSlider.h]: Done

File: components/Slider/src/MDCSlider.m (hunk 5 above)
Comment (romoore) [reply components/Slider/src/MDCSlider.m]: Done

**Review by ianegordon — APPROVED** [review]

**Comment by romoore** [comment]
How do I get rid of the "WIP" check?

**Comment by romoore** [comment]
UGH! WIP will check the commit message. So if a single commit was a work-in-progress, the entire pull gets blocked.

**Pull request merged**